src/inference/support/hmc_core.jl

```diff
@@ -232,53 +234,60 @@ function find_good_eps(model, spl::Sampler{T}, vi::VarInfo) where T
     H_func = gen_H_func()
     θ = vi[spl]
     ϵ = _find_good_eps(θ, lj_func, grad_func, H_func, momentum_sampler)
-    @info "\r[$T] found initial ϵ: $ϵ"
+    @info "[Turing] found initial ϵ: $ϵ"
     return ϵ
 end
 
-function _find_good_eps(θ, lj_func, grad_func, H_func, momentum_sampler; max_num_iters=12)
+function _find_good_eps(θ, lj_func, grad_func, H_func, momentum_sampler; max_num_iters=100)
     @info "[Turing] looking for good initial eps..."
-    ϵ = 0.1
+    ϵ_prime = ϵ = 0.1
+    a_min, a_cross, a_max = 0.25, 0.5, 0.75
+    d = 2.0
 
     p = momentum_sampler()
     H0 = H_func(θ, p, lj_func(θ))
 
     θ_prime, p_prime, τ = _leapfrog(θ, p, 1, ϵ, grad_func)
     h = τ == 0 ? Inf : H_func(θ_prime, p_prime, lj_func(θ_prime))
 
-    delta_H = H0 - h
-    direction = delta_H > log(0.8) ? 1 : -1
-
-    iter_num = 1
+    delta_H = H0 - h # logp(θ`) - logp(θ)
+    direction = delta_H > log(a_cross) ? 1 : -1
 
     # Heuristically find optimal ϵ
-    while (iter_num <= max_num_iters)
-        θ = θ_prime
-
-        p = momentum_sampler()
-        H0 = H_func(θ, p, lj_func(θ))
-
-        θ_prime, p_prime, τ = _leapfrog(θ, p, 1, ϵ, grad_func)
+    for _ = 1:max_num_iters
+        ϵ_prime = direction == 1 ? d * ϵ : 1/d * ϵ
+        θ_prime, p_prime, τ = _leapfrog(θ, p, 1, ϵ_prime, grad_func)
         h = τ == 0 ? Inf : H_func(θ_prime, p_prime, lj_func(θ_prime))
         Turing.DEBUG && @debug "direction = $direction, h = $h"
 
         delta_H = H0 - h
 
-        if ((direction == 1) && !(delta_H > log(0.8)))
+        Turing.DEBUG && @debug "[Turing] ϵ = $ϵ_prime, accept ratio a = $(min(1,(exp(delta_H))))"
+        if ((direction == 1) && !(delta_H > log(a_cross)))
             break
-        elseif ((direction == -1) && !(delta_H < log(0.8)))
+        elseif ((direction == -1) && !(delta_H < log(a_cross)))
             break
         else
-            ϵ = direction == 1 ? 2.0 * ϵ : 0.5 * ϵ
+            ϵ = ϵ_prime
         end
-
-        iter_num += 1
     end
 
-    while h == Inf  # revert if the last change is too big
-        ϵ = ϵ / 2               # safe is more important than large
-        θ_prime, p_prime, τ = _leapfrog(θ, p, 1, ϵ, grad_func)
+    ϵ, ϵ_prime = ϵ < ϵ_prime ? (ϵ, ϵ_prime) : (ϵ_prime, ϵ) # Ensure ϵ < ϵ_prime
+    for _ = 1:max_num_iters
```

Comment:
Is this doing a binary search to find the largest `ϵ` between `ϵ` and `ϵ_prime` which doesn't throw numerical error?

Reply:
sort of - this step ensures the final accept ratio is in the interval `[a_min, a_max]`